plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -506,13 +508,53 @@ public FlowExecutionOwner getOwner() {
     private TimingFlowNodeStorage createStorage() throws IOException {
         FlowNodeStorage wrappedStorage;
 
-        FlowDurabilityHint hint = getDurabilityHint();
-        wrappedStorage = (hint.isPersistWithEveryStep())
-                ? new SimpleXStreamFlowNodeStorage(this, getStorageDir())
-                : new BulkFlowNodeStorage(this, getStorageDir());
+        if (this.storageDir != null && this.storageDir.endsWith("-completed")) {
+           wrappedStorage = new BulkFlowNodeStorage(this, getStorageDir());
+        } else {
+            FlowDurabilityHint hint = getDurabilityHint();
+            wrappedStorage = (hint.isPersistWithEveryStep())
+                    ? new SimpleXStreamFlowNodeStorage(this, getStorageDir())
+                    : new BulkFlowNodeStorage(this, getStorageDir());
+        }
         return new TimingFlowNodeStorage(wrappedStorage);
     }
 
+    /**
+     * Called when the build completes to migrate from {@link SimpleXStreamFlowNodeStorage} to
+     * {@link BulkFlowNodeStorage} to improve read performance for completed builds.
+     */
+    private void optimizeStorage(FlowNode flowEndNode) {
+        if (storage.delegate instanceof SimpleXStreamFlowNodeStorage) {
+            String newStorageDir = (this.storageDir != null) ? this.storageDir + "-completed" : "workflow-completed";
+            try {
+                FlowNodeStorage newStorage = new BulkFlowNodeStorage(this, new File(this.owner.getRootDir(), newStorageDir));
+                DepthFirstScanner scanner = new DepthFirstScanner();
+                scanner.setup(flowEndNode);
+                // The hope is that by doing this right when the execution completes, most of the nodes will already be
+                // cached in memory, saving us the cost of having to read them all from disk.
+                for (FlowNode node : scanner) {
+                    newStorage.storeNode(node, true);
+                }
+                newStorage.flush();
+                File oldStorageDir = getStorageDir();
+                synchronized (this) {
+                    this.storageDir = newStorageDir;
+                    // TODO: A more conservative option could be to instead keep using the current storage until after
+                    // a restart. In that case we'd have to defer deletion of the old storage dir and handle it in
+                    // initializeStorage.
+                    this.storage = new TimingFlowNodeStorage(newStorage);
```

Comment:
Not sure if this is safe or not.

Reply:
I think there is some potential for race conditions here. Methods like `CpsFlowExecution.getNode` are not `synchronized`, so something could call `getNode` while `storage` still points to `SimpleXStreamFlowNodeStorage`, and then if that node is not in the cache, it will be read from disk, but in the meantime `optimizeStorage` could be  deleting `workflow/` after swapping `this.storage`, leading to `getNode` temporarily returning `null`, or the `workflow/` deletion failing, etc.

I can think of a few options:
1. Only delete `workflow/` in `initializeStorage` if `this.storageDir` is `workflow-completed`. Not ideal if the build never actually gets looked at again, because `workflow/` will contain redundant data, but simple.
2. Use `Cleaner` in `optimizeStorage` to register an action to delete the old storage directory once the old `this.storage.delegate` is phantom reachable instead of doing it immediately. Should work fine in the happy path, but if Jenkins crashes or something `workflow/` might not get cleaned up.
3. Add some kind of locking around `this.storage`. Has potential for bugs and general performance issues. Perhaps a `ReadWriteLock` would perform well enough.

Reply:
The third option seems most straightforward.

Reply:
Yes I thought so as well, but it got quite messy. See 4b89fed3df71d639e4f7e3959d9ff26b1f488d1c. It is awkward because although many methods that consume `storage` probably cannot ever run concurrently with `onProgramEnd` and so do not need any locking, it is not immediately obvious and so I am not really comfortable with skipping the locks selectively. Adding a new lock also introduces the possibility of deadlock if something attempts to acquire the monitor for `CpsFlowExecution` while holding `storageLock` and the build completes concurrently (nothing does this today as far as I can see).

The `Cleaner` approach is quite a bit simpler in terms of code changes, for what that's worth. Perhaps we could combine that approach with deletion of `workflow/` in `initializeStorage` to guarantee deletion in most cases without having to introduce the new lock and change all associated code.

Reply:
Hmm, maybe actually I can reuse the lock inside of `TimingFlowNodeStorage` instead to simplify things.

Reply:
Yes I think [62119b8](https://github.com/jenkinsci/workflow-cps-plugin/pull/807/commits/62119b8e009ee491f64f008e56ea85bbd9c743fd) is safe and it is far simpler than introducing another lock. I wish I had noticed it yesterday...